sale_attached_product/models/sale_order.py

```diff
@@ -0,0 +1,92 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_attached_lines(self):
+        self.ensure_one()
+        return self.order_line.filtered("is_attached_line")
+
+    def _get_attached_line_values_product(self, line, product):
+        """Create attached line"""
+
+        def _execute_onchanges(records, field_name):
+            """Helper methods that executes all onchanges associated to a field."""
+            for onchange in records._onchange_methods.get(field_name, []):
+                for record in records:
+                    onchange(record)
+
+        # We prepare a new line and trigger the proper onchanges to ensure we get the
+        # right line values (price unit according to the customer pricelist, taxes, ect)
+        order_line = self.order_line.new(
+            {"order_id": self.id, "product_id": product.id}
+        )
+        _execute_onchanges(order_line, "product_id")
+        order_line.update({"product_uom_qty": line.product_uom_qty})
+        _execute_onchanges(order_line, "product_uom_qty")
+        vals = order_line._convert_to_write(order_line._cache)
+        vals.update({"is_attached_line": True, "attached_from_line_id": line.id})
+        return vals
+
+    def _create_attached_line(self, lines):
+        vals_list = []
+        for line in lines:
+            for product in line.product_id.product_tmpl_id.attached_product_ids:
+                vals_list.append(self._get_attached_line_values_product(line, product))
+        if vals_list:
+            self.write({"order_line": [(0, False, value) for value in vals_list]})
+
+    def _remove_attached_lines(self):
+        self.ensure_one()
+        lines = self._get_attached_lines()
+        lines.filtered(lambda x: not x.attached_from_line_id).unlink()
+
+    def _create_attached_lines(self):
+        self.ensure_one()
+        self._create_attached_line(
+            self.order_line.filtered(
+                lambda x: not x.attached_line_ids
+                and x.product_id.product_tmpl_id.attached_product_ids
+            )
+        )
+
+    def _update_attached_lines(self):
+        self.ensure_one()
+        lines_with_attachements = self.order_line.filtered("attached_line_ids")
+        # TODO: Improve this
+        for line in lines_with_attachements:
+            attached_line_qtys = set(line.attached_line_ids.mapped("product_uom_qty"))
+            if any(q != line.product_uom_qty for q in attached_line_qtys):
+                line.attached_line_ids.update({"product_uom_qty": line.product_uom_qty})
+
+    def recompute_attached_products(self):
+        """Recurrent method for recomputing attached lines"""
+        for order in self.filtered(lambda x: x.state not in {"done", "cancel"}):
+            order._remove_attached_lines()
+            order._create_attached_lines()
+            order._update_attached_lines()
```

Comment:
I'm not sure of that design. 
If I understand correctly, on each write on a sale order (adding / changing a line for exemple), it will drop + create + update all the attached lines.
With the current design, If 
1 I had a main product that adds an attached product, 
2 I change any information on the attached line. (description / price / discount / etc...)
3 I update the sale order
-> the changed I did will be dropped don't you think ?

What about to have a lighter algorithm like 
```_check_attached_lines```.
-> if lines with main product doesn't have attached lines -> create it
-> if lines with attaches lines are updated check if value (product_id / product_uom_qty) are the same. If not update the attached lines.
-> regarding the deletion of main lines, maybe adding ondelete cascade on attached_from_line_id could do the work. Don't you think ?

We can though keep a button "``button_reset_attached_lines`` to recreate all the lines.